packages/contractkit/src/identity/offchain-data-wrapper.ts

```diff
@@ -74,46 +89,93 @@ export default class OffchainDataWrapper {
 
   readDataFrom = makeAsyncThrowable(this.readDataFromAsResult.bind(this))
 
-  async writeDataTo(data: string, dataPath: string) {
+  async writeDataTo(
+    data: Buffer,
+    signature: string,
+    dataPath: string
+  ): Promise<OffchainErrors | void> {
     if (this.storageWriter === undefined) {
-      throw new Error('no storage writer')
+      return new NoStorageProvider()
+    }
+
+    try {
+      await Promise.all([
+        this.storageWriter.write(data, dataPath),
+        await this.storageWriter.write(
+          Buffer.from(trimLeading0x(signature), 'hex'),
+          dataPath + '.signature'
+        ),
+      ])
+    } catch (e) {
+      return new FetchError()
     }
-    await this.storageWriter.write(data, dataPath)
-    // TODO: Prefix signing abstraction
-    const sig = await NativeSigner(this.kit.web3.eth.sign, this.self).sign(data)
-    await this.storageWriter.write(sig, dataPath + '.signature')
   }
 }
 
 class StorageRoot {
-  constructor(readonly account: string, readonly address: string) {}
-
-  // TODO: Add decryption metadata (i.e. indicates ciphertext to be decrypted/which key to use)
-  async readAsResult(dataPath: string): Promise<Result<string, OffchainErrors>> {
-    const data = await fetch(this.address + dataPath)
-    if (!data.ok) {
+  constructor(
+    readonly wrapper: OffchainDataWrapper,
+    readonly account: string,
+    readonly address: string
+  ) {}
+
+  async readAndVerifySignature(
+    dataPath: string,
+    buildTypedData: (x: Buffer) => Promise<EIP712TypedData>
+  ): Promise<Result<Buffer, OffchainErrors>> {
+    let dataResponse, signatureResponse
+
+    try {
+      ;[dataResponse, signatureResponse] = await Promise.all([
+        fetch(this.address + dataPath),
+        fetch(this.address + dataPath + '.signature'),
```

Comment:
We should use `url.resolve` to concatenate the data path with the URL root instead of simple concatenation. https://nodejs.org/dist/latest-v8.x/docs/api/url.html#url_url_resolve_from_to